components/tikv_util/src/lru.rs

```diff
@@ -132,90 +133,131 @@ impl<K> Trace<K> {
     }
 }
 
-pub struct LruCache<K, V> {
-    map: HashMap<K, ValueEntry<K, V>>,
-    trace: Trace<K>,
-    capacity: usize,
+pub trait CacheSizeTrace<K, V>: Default {
+    fn current_size(&self) -> u64;
+    fn trace_insert(&mut self, key: &K, value: &V);
+    fn trace_remove(&mut self, key: &K, value: &V);
+    fn set(&mut self, val: u64);
 }
 
-impl<K, V> LruCache<K, V> {
-    pub fn with_capacity(capacity: usize) -> LruCache<K, V> {
-        LruCache::with_capacity_and_sample(capacity, 0)
+pub struct CountTrace<K, V>(pub u64, PhantomData<(K, V)>);
+
+impl<K, V> CacheSizeTrace<K, V> for CountTrace<K, V> {
+    fn current_size(&self) -> u64 {
+        self.0
     }
 
-    pub fn with_capacity_and_sample(mut capacity: usize, sample_mask: usize) -> LruCache<K, V> {
-        if capacity == 0 {
-            capacity = 1;
-        }
-        LruCache {
-            map: HashMap::default(),
-            trace: Trace::new(sample_mask),
-            capacity,
-        }
+    fn trace_insert(&mut self, _: &K, _: &V) {
+        self.0 += 1;
+    }
+
+    fn trace_remove(&mut self, _: &K, _: &V) {
+        self.0 -= 1;
+    }
+
+    fn set(&mut self, val: u64) {
+        self.0 = val;
+    }
+}
+
+impl<K, V> Default for CountTrace<K, V> {
+    fn default() -> Self {
+        Self(0, PhantomData::default())
     }
+}
+
+pub struct RawLruCache<K, V, T> {
```

Comment:
`PhantomData` is unnecessary and we can replace LruCache type alias by [_Default Generic Type Parameters_](https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#default-generic-type-parameters-and-operator-overloading).
```suggestion
pub struct CountTrace<K, V>(pub u64);

impl<K, V> CacheSizeTrace<K, V> for CountTrace<K, V> {
    fn current_size(&self) -> u64 {
        self.0
    }

    fn trace_insert(&mut self, _: &K, _: &V) {
        self.0 += 1;
    }

    fn trace_remove(&mut self, _: &K, _: &V) {
        self.0 -= 1;
    }

    fn set(&mut self, val: u64) {
        self.0 = val;
    }
}

impl Default for CountTrace {
    fn default() -> Self {
        Self(0)
    }
}

pub struct LruCache<K, V, T: CacheSizeTrace<K, V> = CountTrace> {
```

Reply:
Default generic type may not work very well because we still need to indicate the type parameter explicitly when constructing an object or implementing some methods.
IMO defining a `Raw` type to distinguish the default manner and the custom manner is a good pattern.